src/bin/pg_autoctl/supervisor.c

```diff
@@ -612,9 +959,140 @@ supervisor_shutdown_sequence(Supervisor *supervisor)
 }
 
 
+/*
+ * supervisor_handle_dynamic calls the dynamic handler if exits. It is the
+ * handler's responsibility to add or remove dynamic services using the provided
+ * api.
+ *
+ * Returns the net number of successfully added and removed services. This
+ * number can be negative if more services have been removed than added.
+ */
+static int
+supervisor_handle_dynamic(Supervisor *supervisor)
+{
+	int numChanged = 0;
+
+	/*
+	 * here we call the dyn function if defined, which should return the
+	 * diff of number of services affected, e.g. 1 started => 1, 1 stopped =>
+	 * -1, 2 started and 3 stopped => -1 etc
+	 */
+	if (supervisor->dynamicHandler)
+	{
+		numChanged = supervisor->dynamicHandler(supervisor,
+												supervisor->dynamicHandlerArg);
+	}
+
+	asked_to_handle_dynamic = 0;
+	return numChanged;
+}
+
+
+/*
+ * supervisor_handle_failed_restart decides what action to take when a service
+ * has failed to restart. Currently one of the three can happen:
+ *	* continue running the rest of the services
+ *	* exit the whole program with a happy code
+ *	* exit the whole program with failed code
+ *
+ * The action is depended on whether the service is held in the dynamic array.
+ *
+ * The function can be greatly simplified. It is intentionally left a bit
+ * verbose for the benefit of readability.
+ */
+static void
+supervisor_handle_failed_restart(Supervisor *supervisor, Service *service,
+								 int status)
+{
+	int returnCode = WEXITSTATUS(status);
+	bool isDynamic = false;
+
+
+	/*
+	 * If we are already going handling a shutdown, then do not update it
+	 */
+	if (supervisor->shutdownSequenceInProgress)
+	{
+		log_trace("supervisor_handle_failed_restart: shutdownSequenceInProgress");
+		return;
+	}
+
+	/*
+	 * Loop through the dynamic service array and if it found, then mark it
+	 */
```

Comment:
I am pretty sure we should have (or already have) an internal API to take care of that operation.